firebase-firestore/src/main/java/com/google/firebase/firestore/core/TransactionRunner.java

```diff
@@ -0,0 +1,119 @@
+  private AsyncQueue asyncQueue;
+  private RemoteStore remoteStore;
+  private Function<Transaction, Task<TResult>> updateFunction;
+  private int retries;
+
+  private ExponentialBackoff backoff;
+  private TaskCompletionSource<TResult> taskSource = new TaskCompletionSource<>();
+
+  public TransactionRunner(
+      AsyncQueue asyncQueue,
+      RemoteStore remoteStore,
+      Function<Transaction, Task<TResult>> updateFunction,
+      int retries) {
+    hardAssert(retries >= 0, "Got negative number of retries for transaction.");
+
+    this.asyncQueue = asyncQueue;
+    this.remoteStore = remoteStore;
+    this.updateFunction = updateFunction;
+    this.retries = retries;
+
+    backoff = new ExponentialBackoff(asyncQueue, TimerId.RETRY_TRANSACTION);
+  }
+
+  /** Runs the transaction and sets the result in taskSource. */
+  public void runTransaction() {
+    backoff.backoffAndRun(
+        () -> {
+          final Transaction transaction = remoteStore.createTransaction();
+          updateFunction
+              .apply(transaction)
+              .addOnCompleteListener(
+                  asyncQueue.getExecutor(),
+                  new OnCompleteListener<TResult>() {
+                    @Override
+                    public void onComplete(@NonNull Task<TResult> userTask) {
+                      if (!userTask.isSuccessful()) {
+                        if (retries > 0 && isRetryableTransactionError(userTask.getException())) {
+                          retries -= 1;
+                          runTransaction();
+                        } else {
+                          taskSource.setException(userTask.getException());
+                        }
+                      } else {
+                        transaction
+                            .commit()
+                            .addOnCompleteListener(
+                                asyncQueue.getExecutor(),
+                                new OnCompleteListener<Void>() {
+                                  @Override
+                                  public void onComplete(@NonNull Task<Void> commitTask) {
+                                    if (commitTask.isSuccessful()) {
+                                      taskSource.setResult(userTask.getResult());
+                                    } else if (retries > 0
+                                        && isRetryableTransactionError(commitTask.getException())) {
+                                      retries -= 1;
+                                      runTransaction();
+                                    } else {
+                                      taskSource.setException(commitTask.getException());
+                                    }
+                                  }
+                                });
+                      }
+                    }
+                  });
+        });
+  }
+
+  /** Returns the result of the transaction after it has been run. */
+  public Task<TResult> getTask() {
```

Comment:
Is there a reason for this to be a separate function?  I think I'd just make runTransaction() return the `Task<TResult>`.

This will also probably port better to iOS (`runTransaction()` will just accept a completion callback instead of returning a Task).